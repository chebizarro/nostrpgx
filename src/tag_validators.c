#include "postgres.h"
#include "fmgr.h"
#include "utils/builtins.h"
#include "executor/spi.h"
#include "utils/jsonb.h"
#include "catalog/pg_type.h"
#include "utils/lsyscache.h"
#include "utils/elog.h"
#include "tag_validators.h"

typedef void (*tag_validator)(const char *event_id, const char *tag_value, Jsonb *metadata);

static void validate_e_tag(const char *event_id, const char *tag_value, Jsonb *metadata);
static void validate_p_tag(const char *event_id, const char *tag_value, Jsonb *metadata);

void register_tag_validators(HTAB *tag_validator_map) {
    TagValidatorEntry *entry;

    entry = hash_search(tag_validator_map, "e", HASH_ENTER, NULL);
    entry->validator = validate_e_tag;

    entry = hash_search(tag_validator_map, "p", HASH_ENTER, NULL);
    entry->validator = validate_p_tag;

    // Add more validators here
}

static void validate_e_tag(const char *event_id, const char *tag_value, Jsonb *metadata) {
    char sql[256];
    snprintf(sql, sizeof(sql), "SELECT 1 FROM events WHERE event_id = '%s'", tag_value);
    if (SPI_exec(sql, 0) == SPI_OK_SELECT) {
        if (SPI_processed == 0) {
            ereport(ERROR, (errmsg("Invalid event ID reference in 'e' tag: %s", tag_value)));
        }
    }

    // Additional operation: Insert into another table
    snprintf(sql, sizeof(sql), "INSERT INTO event_references (event_id, referenced_event_id) VALUES ('%s', '%s')", event_id, tag_value);
    SPI_exec(sql, 0);
}

static void validate_p_tag(const char *event_id, const char *tag_value, Jsonb *metadata) {
    if (strlen(tag_value) != 64) {
        ereport(ERROR, (errmsg("Invalid pubkey format in 'p' tag: %s", tag_value)));
    }

    // Additional operation: Create an index
    char sql[256];
    snprintf(sql, sizeof(sql), "CREATE INDEX IF NOT EXISTS idx_pubkey_%s ON events (author)", tag_value);
    SPI_exec(sql, 0);
}

void validate_default_tag(const char *event_id, const char *tag_value, Jsonb *metadata) {
    ereport(NOTICE, (errmsg("Inserting unknown tag: %s", tag_value)));
}
