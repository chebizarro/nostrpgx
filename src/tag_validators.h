#ifndef TAG_VALIDATORS_H
#define TAG_VALIDATORS_H

#include "utils/hsearch.h"

typedef void (*tag_validator)(const char *event_id, const char *tag_value, Jsonb *metadata);

typedef struct {
    char tag_name[16];
    tag_validator validator;
} TagValidatorEntry;

void register_tag_validators(HTAB *tag_validator_map);
void validate_default_tag(const char *event_id, const char *tag_value, Jsonb *metadata);


#endif // TAG_VALIDATORS_H
