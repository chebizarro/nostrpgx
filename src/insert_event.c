/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "postgres.h"
#include "fmgr.h"
#include "utils/builtins.h"
#include "executor/spi.h"
#include "utils/jsonb.h"
#include "catalog/pg_type.h"
#include "utils/lsyscache.h"
#include "utils/elog.h"
#include "tag_validators.h"
#include <inttypes.h> // Include inttypes.h for PRIi64

extern HTAB *tag_validator_map;

Datum insert_event(PG_FUNCTION_ARGS);
static void validate_and_index_tag(const char *event_id, const char *tag_name, const char *tag_value, Jsonb *metadata);

Datum insert_event(PG_FUNCTION_ARGS) {
    Jsonb *event_jsonb = PG_GETARG_JSONB_P(0);
    JsonbIterator *it;
    JsonbValue v;
    JsonbIteratorToken type;

    const char *event_id = NULL;
    const char *author = NULL;
    int kind = 0;
    int64 created_at = 0;
    const char *content = NULL;
    Jsonb *tags_jsonb = NULL;

    it = JsonbIteratorInit(&event_jsonb->root);

    while ((type = JsonbIteratorNext(&it, &v, false)) != WJB_DONE) {
        if (type == WJB_KEY) {
            if (strcmp(v.val.string.val, "event_id") == 0) {
                JsonbIteratorNext(&it, &v, false);
                event_id = v.val.string.val;
            } else if (strcmp(v.val.string.val, "author") == 0) {
                JsonbIteratorNext(&it, &v, false);
                author = v.val.string.val;
            } else if (strcmp(v.val.string.val, "kind") == 0) {
                JsonbIteratorNext(&it, &v, false);
                kind = DatumGetInt32(DirectFunctionCall1(int4in, CStringGetDatum(v.val.string.val)));
            } else if (strcmp(v.val.string.val, "created_at") == 0) {
                JsonbIteratorNext(&it, &v, false);
                created_at = DatumGetInt64(DirectFunctionCall1(int8in, CStringGetDatum(v.val.string.val)));
            } else if (strcmp(v.val.string.val, "content") == 0) {
                JsonbIteratorNext(&it, &v, false);
                content = v.val.string.val;
            } else if (strcmp(v.val.string.val, "tags") == 0) {
                JsonbIteratorNext(&it, &v, false);
                tags_jsonb = JsonbValueToJsonb(&v);
            }
        }
    }

    if (event_id == NULL || author == NULL || content == NULL || tags_jsonb == NULL) {
        ereport(ERROR, (errmsg("Invalid event JSON")));
    }

    SPI_connect();

    char insert_event_sql[1024];
    snprintf(insert_event_sql, sizeof(insert_event_sql),
             "INSERT INTO events (event_id, author, kind, created_at, content) "
             "VALUES ('%s', '%s', %d, %" PRIi64 ", '%s')",
             event_id, author, kind, created_at, content);

    SPI_exec(insert_event_sql, 0);

    JsonbIterator *tags_it = JsonbIteratorInit(&tags_jsonb->root);
    JsonbValue tag;
    while ((type = JsonbIteratorNext(&tags_it, &tag, false)) != WJB_DONE) {
        if (type == WJB_ELEM) {
            if (tag.type == jbvArray) {
                const char *tag_name = NULL;
                const char *tag_value = NULL;
                JsonbParseState *state = NULL;
                JsonbValue *metadata = NULL;

                JsonbIterator *tag_it = JsonbIteratorInit(tag.val.binary.data);
                JsonbValue tag_elem;
                int tag_elem_index = 0;

                while ((type = JsonbIteratorNext(&tag_it, &tag_elem, false)) != WJB_DONE) {
                    if (tag_elem.type == jbvString) {
                        if (tag_elem_index == 0) {
                            tag_name = tag_elem.val.string.val;
                        } else if (tag_elem_index == 1) {
                            tag_value = tag_elem.val.string.val;
                        } else {
                            if (metadata == NULL) {
                                metadata = pushJsonbValue(&state, WJB_BEGIN_ARRAY, NULL);
                            }
                            metadata = pushJsonbValue(&state, WJB_ELEM, &tag_elem);
                        }
                        tag_elem_index++;
                    }
                }

                if (metadata != NULL) {
                    metadata = pushJsonbValue(&state, WJB_END_ARRAY, NULL);
                }

                if (tag_name == NULL || tag_value == NULL) {
                    ereport(ERROR, (errmsg("Invalid tag format")));
                }

                Jsonb *metadata_jsonb = metadata != NULL ? JsonbValueToJsonb(metadata) : NULL;

                validate_and_index_tag(event_id, tag_name, tag_value, metadata_jsonb);

                char *metadata_cstring = metadata_jsonb != NULL ? JsonbToCString(NULL, &metadata_jsonb->root, VARSIZE(metadata_jsonb)) : "";

                char insert_tag_sql[1024];
                snprintf(insert_tag_sql, sizeof(insert_tag_sql),
                         "INSERT INTO tags (event_id, tag_name, tag_value, metadata) "
                         "VALUES ('%s', '%s', '%s', '%s')",
                         event_id, tag_name, tag_value, metadata_cstring);

                SPI_exec(insert_tag_sql, 0);

                if (metadata_jsonb != NULL) {
                    pfree(metadata_cstring);
                }
            }
        }
    }

    SPI_finish();

    PG_RETURN_VOID();
}

static void validate_and_index_tag(const char *event_id, const char *tag_name, const char *tag_value, Jsonb *metadata) {
    TagValidatorEntry *entry = hash_search(tag_validator_map, tag_name, HASH_FIND, NULL);
    if (entry) {
        entry->validator(event_id, tag_value, metadata);
    } else {
        validate_default_tag(event_id, tag_value, metadata);
    }
}
