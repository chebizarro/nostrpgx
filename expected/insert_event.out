insert_event 
--------------
 
(1 row)

insert_event 
--------------
 
(1 row)
